### PR TITLE
Plugable coins API

### LINKED ISSUE #2155: ERC20 Gas limit & Gas price Withdrawal bug
In Peatio (v2) admin panel we have Gas price and Gas limit for transaction fees in Ethereum BC, its working fine with ETH it self, but with ERC20 withdrawal you will have an issue with insufficient funds a lot.

STR:
1. Set up ERC20 currency with expected and sufficient gas price per transaction  
2. Set up Fee wallet in ETH and feed it with coins to pay fees
3. ERC20 Deposit/Hot/Warm wallet with Maximum Balance 0.0/0.5/1.0 respectively (Main idea to get deposit amount split between wallets, so check your Hot and Warm wallet for current balance) 
4. Make a deposit to a user deposit address in amount of 0.6+ 

Expected:
Now we should receive to user Deposit wallet amount that is equal to our Gas limit*Gas price in ETH from fee wallet, after what withdraw of ERC20 should happen to Hot and Warm wallets.

Actual:
We do receive Gas limit*Gas price in ETH from fee wallet, but the Collection process will not be continued, our deposit will be Skipped.

Reason:
The Fee we are expecting to use in ERC20 transaction are meant only for one withdrawal, instead we have splitted our deposit into two withdrawal operations and fee amount taken from the fee wallet will not be sufficient.

### PR DESCRIPTION
**Pull request by ysv**
Finalization todo list: 
 

- [x] Revert Wallet GATEWAY validation and usage (@mnaichuk)
- [x] Remove BlockchainAddress, use Hash instead (@mnaichuk)
- [x] Finish WalletService2 specs (@mnaichuk)
- [x] https://github.com/rubykube/peatio/pull/2168/files#r275769063 (@mnaichuk)
- [x] Build Exception hierarchy for Blockchain & Wallet
- [x] Add Peato::Blockchain::Abstact #load_balance
  - [x] Add to Bitcoin
  - [x] Add to Ethereum (@mnaichuk)
- [x] Replace Blockchain #blockchain_api with v2
- [x] Remove all currencies except btc eth & erc20 from specs
- [x] Implement Ethereum::Wallet adapter
- [x] Integrate substract_fee option into bitcoin with specs(@mnaichuk)
- [x] Update daemons to use WalletService2 & BlockchainService2
  - [x] Blockchain daemon with specs(@mnaichuk)
  - [x] Withdraw daemons with specs
- [x] Finalize TODOs
  - [x] Remove unused resources/
  - [x] Finish TODOs in specs
  - [x] Uncomment wallet validation 
- [x] Move core code to gem
- [x] Functional testing in microkube
  - [x] Without plugins
  - [x] With plugin
  - [x] Final fixes
- [x] Finalize
  - [x] Release peatio 0.5.2 and 
  - [x] Update peatio to 0.5.2 in peatio-litecoin
  - [x] Release peatio-litecoin
  - [x] Upeate peatio & peatio-litecoin gems in pr

closed #2045 
closes #2155

### CODE REVIEW

File: spec/services/blockchain_service2_spec.rb [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+                          txid: 'fake_txid',
+                          block_number: 0,
+                          txout: 4,
+                          type: Deposits::Coin)
+        end
+        before do
+          fake_adapter.stubs(:fetch_block!).returns([transaction])
+          service.process_block(block_number)
+        end
+        it { expect(Deposits::Coin.find_by(txid: transaction.hash).block_number).to eq(transaction.block_number) }
```
Comment (ysv) [inline spec/services/blockchain_service2_spec.rb]: @mnaichuk check that message to deposit_collection_fees was published

File: spec/services/blockchain_service2_spec.rb [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+        fake_adapter.stubs(:fetch_block!).returns(expected_transactions)
+        service.process_block(block_number)
+      end
+
+      subject { Deposits::Coin.where(currency: [fake_currency1, fake_currency2]) }
+
+      it { expect(subject.count).to eq 2 }
+
+      it 'create for two currency' do
+        expect(Deposits::Coin.where(currency: fake_currency1).count).to eq 1
```
Comment (ysv) [inline spec/services/blockchain_service2_spec.rb]: `Deposits::Coin.exists?(currency: fake_currency1)`

File: spec/services/blockchain_service2_spec.rb [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+    end
+
+    subject { Withdraws::Coin.where(currency: [fake_currency1, fake_currency2]) }
+
+    it { expect(subject.find_by(txid: expected_transactions.first.hash).block_number).to eq(expected_transactions.first.block_number) }
+
+    it { expect(subject.find_by(txid: expected_transactions.third.hash).block_number).to eq(expected_transactions.third.block_number) }
+  end
+
+  context 'single withdrawal was succed during block processing' do
```
Comment (ysv) [inline spec/services/blockchain_service2_spec.rb]: succeed

File: spec/services/blockchain_service2_spec.rb [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+
+    before do
+      fake_adapter.stubs(:latest_block_number).returns(10)
+      fake_adapter.stubs(:fetch_block!).returns(expected_transactions)
+      service.process_block(block_number)
+    end
+
+    it { expect(withdrawal.reload.succeed?).to eq(true) }
+  end
+end
```
Comment (ysv) [inline spec/services/blockchain_service2_spec.rb]: add specs where we process blocks one by one

File: app/models/deposit.rb [hunk 5]
```diff
@@ -85,13 +85,7 @@ def plus_funds
...
 
   def collect!(collect_fee = true)
-    if coin?
-      if currency.is_erc20? && collect_fee
-        AMQPQueue.enqueue(:deposit_collection_fees, id: id)
-      else
-        AMQPQueue.enqueue(:deposit_collection, id: id)
-      end
-    end
+    AMQPQueue.enqueue(:deposit_collection_fees, id: id) if coin? && collect_fee
```
Comment (ysv) [inline app/models/deposit.rb]: we still need to publish to `deposit_collection` if collect_fee is false

**Comment by ysv** [comment]
Outdated

File: app/services/blockchain_service.rb [hunk 6]
```diff
@@ -1,116 +1,121 @@
...
 
-  class Base
+  attr_reader :blockchain, :adapter
 
-    attr_reader :blockchain, :client
+  def initialize(blockchian)
+    @blockchain = blockchian
+    @adapter = Peatio::Blockchain.registry[blockchian.client.to_sym]
+    @adapter.configure(server: @blockchain.server,
+                       currencies: @blockchain.currencies.map(&:to_blockchain_api_settings))
```
Comment (mnaichuk) [inline app/services/blockchain_service.rb]: I think we need to add @blockchain.currencies.enabled.map(&:to_blockchain_api_settings)), @ysv ?

File: app/services/wallet_service.rb [hunk 7]
```diff
@@ -1,95 +1,136 @@
...
-      raise Error, "Wrong WalletService name #{wallet_service}"
-    end
+class WalletService
+  attr_reader :wallet, :adapter
+
+  def initialize(wallet)
+    @wallet = wallet
+    @adapter = Peatio::Wallet.registry[wallet.gateway.to_sym]
+    @adapter.configure(wallet: @wallet.to_wallet_api_settings,
+                       currency: @wallet.currency.to_blockchain_api_settings)
```
Comment (mnaichuk) [inline app/services/wallet_service.rb:8]: and enabled here too

**Review by mod — APPROVED** [review]

**Comment by ysv** [comment]
Finalization todo list: 
 

- [x] Revert Wallet GATEWAY validation and usage (@mnaichuk)
- [x] Remove BlockchainAddress, use Hash instead (@mnaichuk)
- [x] Finish WalletService2 specs (@mnaichuk)
- [x] https://github.com/rubykube/peatio/pull/2168/files#r275769063 (@mnaichuk)
- [x] Build Exception hierarchy for Blockchain & Wallet
- [x] Add Peato::Blockchain::Abstact #load_balance
  - [x] Add to Bitcoin
  - [x] Add to Ethereum (@mnaichuk)
- [x] Replace Blockchain #blockchain_api with v2
- [x] Remove all currencies except btc eth & erc20 from specs
- [x] Implement Ethereum::Wallet adapter
- [x] Integrate substract_fee option into bitcoin with specs(@mnaichuk)
- [x] Update daemons to use WalletService2 & BlockchainService2
  - [x] Blockchain daemon with specs(@mnaichuk)
  - [x] Withdraw daemons with specs
- [x] Finalize TODOs
  - [x] Remove unused resources/
  - [x] Finish TODOs in specs
  - [x] Uncomment wallet validation 
- [x] Move core code to gem
- [x] Functional testing in microkube
  - [x] Without plugins
  - [x] With plugin
  - [x] Final fixes
- [x] Finalize
  - [x] Release peatio 0.5.2 and 
  - [x] Update peatio to 0.5.2 in peatio-litecoin
  - [x] Release peatio-litecoin
  - [x] Upeate peatio & peatio-litecoin gems in pr


closed #2045 
closes #2155

**Pull request merged**